### ISSUE TITLE
[py2py3] investigate WMCore/WebTools/RESTFormatter.to_string() return type in py3

### ISSUE THREAD

**Issue by mapellidario** [dmwm/WMCore#10222]
**Impact of the new feature**

It should affect only the internals cherrypy when running WMCore from a py3 environment, [this function](https://github.com/cherrypy/cherrypy/blob/c524a0da73a85ec4dd4ae6b1512c25854b19c8de/cherrypy/_cprequest.py#L775) in particular.

**Is your feature request related to a problem? Please describe.**

In #10117  we modernized ` src/python/WMCore/WebTools/RESTFormatter.py ` as follows

```python
+ from builtins import str, bytes
    def to_string(self, data):
        if isinstance(data, GeneratorType):
            return self.json(data)
        if isinstance(data, dict) or isinstance(data, list):
            return json.dumps(data)
-        return str(data)
+        return bytes(str(data), "utf-8") #  equivalent to `str(data).encode("utf-8")`
```

as discussed in [this comment in PR 10117](https://github.com/dmwm/WMCore/pull/10117#discussion_r555779338). This is based on the assumption that the internals of cherrypy use bytes string in py3 also, which is only guessed by looking at [cherrypy py3 code](https://github.com/cherrypy/cherrypy/blob/c524a0da73a85ec4dd4ae6b1512c25854b19c8de/cherrypy/_cprequest.py#L775), this has not been tested yet.

**Describe the solution you'd like**

We should test this assumption in dev, testbed and/or pre-production.

**Describe alternatives you've considered**

If cherrypy requires `RESTFormatter.to_string() ` to return a bytes string in py3 also, then we are good to go. If it requires unicode string in py3, then we need to change our code.

**Comment by mapellidario** [comment]
Depends on having a py3 environment

**Comment by amaltaro** [comment]
@mapellidario Dario, with the full python3 environment available now, this might be a good candidate to be worked on next and get it out of our radar.

**Comment by amaltaro** [comment]
And in order to have a full validation of this module, we need to wait for CRABServer to migrate to the latest WMCore tag and see how it goes. FYI @belforte (tagging you just to keep you in the loop)

**Comment by belforte** [comment]
I am definitely aware that moving CRAB REST code will be not-trivial and surely require changes to what we are using now in 1.4.6.crab branch where indeed I did some horrible hacking to deal with unicode vs bytes. But on the good side, we do not want py2/3 compatibility there, pure py3 will be fine. And I am counting on @mapellidario to take care, while it is not a priority. Current container can keep running over next months. We will of course need to be very careful in case we decide that it is better (or unavaildable) to change what the REST returns to clients, since original developers followed CMS tradition at the time of not bothering to return proper JSON but sort of "printed python dictionaries".  Slowly and carefully, as they say in Trieste (pian e ben, for Dario).

**Comment by mapellidario** [comment]
The purpose of the PR #10753 was not really to fix anything, i wanted to investigate the effects of that conversion from unicode to bytes. I tried removing it in #10753 and nothing changed in our unittests, while in #10117 we saw some failures that then spawned this issue. I think that during our modernization efforts we added some proper checks around `RESTFormatter.to_string()` that guard us from those failures.

Long story short: I can not replicate the old failures, but since this is an interesting topic for both WMCore and CRAB, then i propose to keep this issue open and closing my PR. (if @amaltaro you do not agree, then you can reopen it!)

**Comment by belforte** [comment]
as you prefer. Once we start testing CRABServer with recern WMCore we will surely open issues here as needed. I am not worried about forgetting !

**Comment by mapellidario** [comment]
We eventually decided that we return "native str" (bytes in py2, unicode in py3), which is consistent with the `return json.dumps(data)` of few lines before, and we close this issue. If CRAB will find problems with this method, then we will open a new issue.
Thanks Stefano for checking in and thanks Alan for the review and merging!
